File: Code-Judger/backend/docker/runners/java-runner/WarmRunner.java
import java.io.*;
import java.net.*;
import java.util.concurrent.TimeUnit;
import java.util.UUID;

public class WarmRunner {

    static class StreamGobbler extends Thread {
        private final InputStream is;
        private final ByteArrayOutputStream bos = new ByteArrayOutputStream();

        StreamGobbler(InputStream is) {
            this.is = is;
        }

        @Override
        public void run() {
            try {
                byte[] buffer = new byte[4096];
                int len;
                while ((len = is.read(buffer)) != -1) {
                    bos.write(buffer, 0, len);
                }
            } catch (IOException e) {
                // Ignore
            }
        }

        byte[] getBytes() {
            return bos.toByteArray();
        }
    }

    private static void deleteDirectory(File directory) {
        File[] allContents = directory.listFiles();
        if (allContents != null) {
            for (File file : allContents) {
                deleteDirectory(file);
            }
        }
        directory.delete();
    }

    public static void main(String[] args) {
        try (ServerSocket serverSocket = new ServerSocket(8080)) {
            System.out.println("Java Warm Runner (Binary Protocol) listening on port 8080...");

            while (true) {
                Socket socket = serverSocket.accept();
                File jobDir = null;
                try (
                    DataInputStream in = new DataInputStream(socket.getInputStream());
                    DataOutputStream out = new DataOutputStream(socket.getOutputStream())
                ) {
                    // Read header
                    int codeLen = in.readInt();
                    int inputLen = in.readInt();
                    int timeoutMs = in.readInt();

                    // Read payloads
                    byte[] codeBytes = new byte[codeLen];
                    in.readFully(codeBytes);
                    byte[] inputBytes = new byte[inputLen];
                    in.readFully(inputBytes);

                    // Create unique directory inside /tmp (world-writable)
                    String jobId = UUID.randomUUID().toString();
                    jobDir = new File("/tmp/job-" + jobId);
                    if (!jobDir.mkdirs()) {
                        throw new IOException("Failed to create job directory");
                    }

                    // Write code to Main.java
                    File sourceFile = new File(jobDir, "Main.java");
                    try (FileOutputStream fos = new FileOutputStream(sourceFile)) {
                        fos.write(codeBytes);
                    }

                    // Compile Main.java
                    ProcessBuilder compileBuilder = new ProcessBuilder("javac", sourceFile.getAbsolutePath());
                    Process compileProc = compileBuilder.start();

                    StreamGobbler compileErrGobbler = new StreamGobbler(compileProc.getErrorStream());
                    compileErrGobbler.start();
                    compileProc.waitFor();
                    compileErrGobbler.join();

                    int compileExit = compileProc.exitValue();
                    if (compileExit != 0) {
                        byte[] errBytes = compileErrGobbler.getBytes();
                        // Write compile error response (exitCode = 1, timedOut = 0, runtime = 0)
                        out.writeInt(1);
                        out.writeByte(0);
                        out.writeInt(0);
                        out.writeInt(0);
                        out.writeInt(errBytes.length);
                        out.write(errBytes);
                        continue;
                    }

                    // Execute Java code
                    ProcessBuilder runBuilder = new ProcessBuilder(
                        "java",
                        "-Xss512k",
                        "-Xmx200m",
                        "-cp",
                        jobDir.getAbsolutePath(),
                        "Main"
                    );
                    
                    long startTime = System.nanoTime();
                    Process runProc = runBuilder.start();

                    // Write stdin
                    if (inputLen > 0) {
                        try (OutputStream stdin = runProc.getOutputStream()) {
                            stdin.write(inputBytes);
                            stdin.flush();
                        }
                    } else {
                        runProc.getOutputStream().close();
                    }

                    // Spin up gobblers to drain stdout/stderr
                    StreamGobbler outGobbler = new StreamGobbler(runProc.getInputStream());
                    StreamGobbler errGobbler = new StreamGobbler(runProc.getErrorStream());
                    outGobbler.start();
                    errGobbler.start();

                    boolean finished = runProc.waitFor(timeoutMs, TimeUnit.MILLISECONDS);
                    long endTime = System.nanoTime();
                    int runtimeMs = (int) ((endTime - startTime) / 1_000_000);

                    int exitCode = 0;
                    boolean timedOut = false;

                    if (!finished) {
                        runProc.destroyForcibly();
                        timedOut = true;
                        exitCode = 124;
                    } else {
                        exitCode = runProc.exitValue();
                    }

                    outGobbler.join();
                    errGobbler.join();

                    byte[] stdoutBytes = outGobbler.getBytes();
                    byte[] stderrBytes = errGobbler.getBytes();

                    if (timedOut) {
                        stderrBytes = "Time Limit Exceeded".getBytes("UTF-8");
                    }

                    // Send response header: exitCode (4B), timedOut (1B), runtimeMs (4B), stdoutLen (4B), stderrLen (4B)
                    out.writeInt(exitCode);
                    out.writeByte(timedOut ? 1 : 0);
                    out.writeInt(runtimeMs);
                    out.writeInt(stdoutBytes.length);
                    out.writeInt(stderrBytes.length);
                    
                    // Send response body
                    out.write(stdoutBytes);
                    out.write(stderrBytes);

                } catch (Exception e) {
                    try {
                        // Return generic execution error
                        ByteArrayOutputStream baos = new ByteArrayOutputStream();
                        PrintStream ps = new PrintStream(baos);
                        e.printStackTrace(ps);
                        byte[] errBytes = baos.toByteArray();
                        
                        DataOutputStream errOut = new DataOutputStream(socket.getOutputStream());
                        errOut.writeInt(1);
                        errOut.writeByte(0);
                        errOut.writeInt(0);
                        errOut.writeInt(0);
                        errOut.writeInt(errBytes.length);
                        errOut.write(errBytes);
                    } catch (Exception ignored) {}
                } finally {
                    if (jobDir != null && jobDir.exists()) {
                        deleteDirectory(jobDir);
                    }
                    try {
                        socket.close();
                    } catch (IOException ignored) {}
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: Code-Judger/backend/docker/runners/python-runner/warm_runner.py
import socket
import struct
import subprocess
import tempfile
import os
import time

def recv_all(sock, length):
    data = b''
    while len(data) < length:
        packet = sock.recv(length - len(data))
        if not packet:
            return None
        data += packet
    return data

def main():
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server.bind(('0.0.0.0', 8080))
    server.listen(5)
    print("Python Warm Runner (Binary Protocol) listening on port 8080...", flush=True)

    while True:
        conn, addr = server.accept()
        try:
            # Read 12-byte header: code_len (4B), input_len (4B), timeout_ms (4B)
            header = recv_all(conn, 12)
            if not header:
                conn.close()
                continue
            
            code_len, input_len, timeout_ms = struct.unpack(">III", header)
            
            # Read code and input bodies
            code_bytes = recv_all(conn, code_len) if code_len > 0 else b''
            input_bytes = recv_all(conn, input_len) if input_len > 0 else b''
            
            code = code_bytes.decode('utf-8')
            input_data = input_bytes.decode('utf-8')

            # Create temporary execution file in /tmp (world-writable)
            with tempfile.NamedTemporaryFile(mode='w', suffix='.py', delete=False, dir='/tmp') as f:
                f.write(code)
                temp_file_path = f.name
            
            os.chmod(temp_file_path, 0o644)
            
            start_time = time.perf_counter()
            try:
                result = subprocess.run(
                    ['python3', temp_file_path],
                    input=input_data,
                    capture_output=True,
                    text=True,
                    timeout=timeout_ms / 1000.0
                )
                stdout = result.stdout
                stderr = result.stderr
                exit_code = result.returncode
                timed_out = False
            except subprocess.TimeoutExpired as e:
                stdout = e.stdout or ""
                stderr = e.stderr or "Time Limit Exceeded"
                exit_code = 124
                timed_out = True
            finally:
                try:
                    os.unlink(temp_file_path)
                except:
                    pass
            
            runtime_ms = int((time.perf_counter() - start_time) * 1000)
            
            # Encode response
            stdout_bytes = stdout.encode('utf-8')
            stderr_bytes = stderr.encode('utf-8')
            
            # Response header: exit_code (4B), timed_out (1B), runtime_ms (4B), stdout_len (4B), stderr_len (4B)
            response_header = struct.pack(
                ">iBIII",
                exit_code,
                1 if timed_out else 0,
                runtime_ms,
                len(stdout_bytes),
                len(stderr_bytes)
            )
            
            conn.sendall(response_header + stdout_bytes + stderr_bytes)
        except Exception as e:
            err_msg = str(e).encode('utf-8')
            # Error response: exit_code=1, timed_out=0, runtime_ms=0, stdout_len=0, stderr_len=err_len
            err_header = struct.pack(">iBIII", 1, 0, 0, 0, len(err_msg))
            try:
                conn.sendall(err_header + err_msg)
            except:
                pass
        finally:
            conn.close()

if __name__ == '__main__':
    main()
